Data Science/SleepDisorderPrediction/feature_extraction.ipynb

```diff
@@ -0,0 +1,338 @@
+    "from scipy.linalg import svd\n",
+    "from statsmodels.robust import mad\n",
+    "import nolds\n",
+    "from scipy.integrate import simps\n",
+    "import pandas as pd\n",
+    "\n",
+    "# Load data\n",
+    "data = pd.read_csv(\"/Users/jasper/Desktop/SIT782/test8.csv\", header=None)\n",
+    "data_1 = data[0][0:299]\n",
+    "ppg_signal = np.array([data_1]).flatten()\n",
+    "\n",
+    "# Constants\n",
+    "sampling_rate = 100  # in Hz\n",
+    "epoch_duration = len(ppg_signal) / sampling_rate  # in seconds\n",
+    "\n",
+    "# Find peaks (simulating pulse peaks)\n",
+    "peaks, _ = find_peaks(ppg_signal)\n",
+    "troughs, _ = find_peaks(-ppg_signal)\n",
+    "\n",
+    "# Peak-to-peak intervals (in seconds)\n",
+    "pp_intervals = np.diff(peaks) / sampling_rate\n",
+    "\n",
+    "# Feature calculation functions\n",
+    "def rms(x):\n",
+    "    return np.sqrt(np.mean(x**2))\n",
+    "\n",
+    "def calculate_sdsd(x):\n",
+    "    return np.std(np.diff(x))\n",
+    "\n",
+    "def calculate_rmssd(x):\n",
+    "    return np.sqrt(np.mean(np.diff(x)**2))\n",
+    "\n",
+    "def calculate_areas(signal, inflection_indices):\n",
+    "    area_sys = simps(signal[inflection_indices[:len(inflection_indices)//2]], dx=1/sampling_rate)\n",
+    "    area_dias = simps(signal[inflection_indices[len(inflection_indices)//2:]], dx=1/sampling_rate)\n",
+    "    return area_sys, area_dias\n",
+    "\n",
+    "def calculate_pulse_width(peaks, troughs, sampling_rate):\n",
+    "    widths = []\n",
+    "    for i in range(len(peaks) - 1):\n",
+    "        if i < len(troughs) and peaks[i] < troughs[i]:\n",
+    "            widths.append((troughs[i] - peaks[i]) / sampling_rate)\n",
+    "    return np.array(widths)\n",
+    "\n",
+    "def inflection_points(signal):\n",
+    "    inflections = []\n",
+    "    for i in range(1, len(signal)-1):\n",
+    "        if (signal[i-1] < signal[i] > signal[i+1]) or (signal[i-1] > signal[i] < signal[i+1]):\n",
+    "            inflections.append(i)\n",
+    "    return inflections\n",
+    "    \n",
+    "def higuchi_fd(x, kmax):\n",
+    "    \"\"\"Calculate Higuchi Fractal Dimension (HFD) of a time series.\"\"\"\n",
+    "    N = len(x)\n",
+    "    Lk = np.zeros(kmax)\n",
+    "\n",
+    "    for k in range(1, kmax + 1):\n",
+    "        Lm = np.zeros((k,))\n",
+    "\n",
+    "        for m in range(0, k):\n",
+    "            Lmk = 0\n",
+    "            for i in range(1, int(np.floor((N - m) / k))):\n",
+    "                Lmk += abs(x[m + i * k] - x[m + (i - 1) * k])\n",
+    "            Lm[m] = (Lmk * (N - 1) / (k * np.floor((N - m) / k)))\n",
+    "\n",
+    "        Lk[k - 1] = np.sum(Lm) / k\n",
+    "\n",
+    "    # Perform linear regression to find the slope (fractal dimension)\n",
+    "    Lk_log = np.log(Lk)\n",
+    "    k_log = np.log(range(1, kmax + 1))\n",
+    "    higuchi, _ = np.polyfit(k_log, Lk_log, 1)\n",
+    "\n",
+    "    return higuchi\n",
```

Comment:
Please add docstrings to each function so that anyone who reads the code can easily understand the use of the function. Refer to [this guide from pandas](https://pandas.pydata.org/docs/development/contributing_docstring.html) and [this guide](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) to see examples of docstring formats.